### PR TITLE
github: annotate build warnings

### PR DESCRIPTION
**Pull request by crazy-max**
* ~needs #362~

example in this PR through integration tests: https://github.com/docker/actions-toolkit/pull/365/files#diff-cc857d138abc3f09c4d212a933d59584a266fe71521fd2cd87327dc037558a43R17

image

___

image

___

image

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ on:
 
 env:
   NODE_VERSION: "20"
-  BUILDX_VERSION: "v0.15.0"
+  BUILDX_VERSION: "https://github.com/crazy-max/buildx.git#metadata-warnings"
```
Comment (crazy-max) [inline .github/workflows/test.yml:18]: remove this change when https://github.com/docker/buildx/pull/2511 released

**Comment by crazy-max** [comment]
example in this PR through integration tests: https://github.com/docker/actions-toolkit/pull/365/files#diff-cc857d138abc3f09c4d212a933d59584a266fe71521fd2cd87327dc037558a43R2

image

@daghack @tonistiigi Looking at the annotations it seems we still have the line number in the details. Recall it was supposed to be removed as this is redundant. Is it an oversight?

Also looking at build command output: https://github.com/docker/actions-toolkit/actions/runs/9523193509/job/26254172537#step:10:808

image

Is it possible to disable checks during build entirely as this is already handled using status warnings in build record? I think for GHA it makes more sense to only rely on GitHub Annotations.

Same goes for warning display in job summary: https://github.com/docker/actions-toolkit/actions/runs/9523193509

image

**Comment by crazy-max** [comment]
Made some changes to print the url for more details and also strip ` (line X)` from `warning.short` that is not needed for GitHub Annotations.

Would like this in PR changes:

image

In workflow summary:

image

Workflow logs:

image

**Comment by crazy-max** [comment]
rebased and updated annotation to match at best `--check` with debug enabled:

image

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@ on:
 
 env:
   NODE_VERSION: "20"
-  BUILDX_VERSION: "v0.15.1"
+  BUILDX_VERSION: "https://github.com/docker/buildx.git#d8c9ebde1fdcf659f1fa3efa6ccc27a28b0f1564"  # https://github.com/docker/buildx/pull/2551
```
Comment (crazy-max) [inline .github/workflows/test.yml:18]: needs follow-up to bump to v0.16.0 when released

File: src/github.ts [hunk 3]
```diff
@@ -328,4 +329,34 @@ export class GitHub {
...
+          endLine = range.end.line;
+        }
+      }
+      core.info(`${message} --- startLine: ${startLine}, endLine: ${endLine}`);
+      core.warning(message, {
+        title: title,
+        file: source,
+        startLine: startLine,
+        endLine: endLine
+      });
```
Comment (crazy-max) [inline src/github.ts:359]: Was looking at showing first and last lines matching warning ranges but GitHub annotations does not seem to display that properly:

image

`core.info` returns the right `startLine` and `endLine` through:

```
ConsistentInstructionCasing: Command 'COPy' should match the case of the command majority (lowercase)
More info: https://docs.docker.com/go/dockerfile/rule/consistent-instruction-casing/ --- startLine: 21, endLine: 23
```

I will just set `startLine` for now and open a thread on GitHub community.

File: src/github.ts [hunk 4]
```diff
@@ -328,4 +329,39 @@ export class GitHub {
...
+  public static async annotateBuildWarnings(source: string, warnings?: Array<VertexWarning>): Promise<void> {
+    (warnings ?? []).forEach(warning => {
+      if (!warning.detail || !warning.short) {
+        return;
+      }
+      const title = warning.detail.map(encoded => atob(encoded)).join(' ');
+      let message = atob(warning.short).replace(/\s\(line \d+\)$/, '');
+      if (warning.url) {
+        // https://github.com/docker/buildx/blob/d8c9ebde1fdcf659f1fa3efa6ccc27a28b0f1564/commands/build.go#L854
+        message += `\nMore info: ${warning.url}`;
```
Comment (tonistiigi) [inline src/github.ts:342]: Wonder if this link can be made clickable by adding some markdown around it.

**Review by tonistiigi — APPROVED** [review]

File: src/github.ts (hunk 4 above)
Comment (crazy-max) [reply src/github.ts:342]: Opened https://github.com/orgs/community/discussions/129897

**Pull request merged**